Make it possible to define joins for log tables using `getWaysToJoinToOtherLogTables`

sgiehl opened the pull request:
In some cases it might be useful to join tables to visitor data using other columns than `idVisit` and `idAction`.

There is already a method `LogTable::getWaysToJoinToOtherLogTables`, which makes it possible to define ways to join the table other than using `idVisit` and `idAction`.
But currently that method is only used in PrivacyManager in order to remove related data.

This change should also make it possible to use tables joined by `getWaysToJoinToOtherLogTables` as dimensions and in segmentation.

sgiehl committed:
Make it possible to define joins for log tables using getWaysToJoinToOtherLogTables

sgiehl committed:
Adds some tests for custom log table joins

sgiehl commented on core/DataAccess/LogQueryBuilder/JoinGenerator.php:
```diff
@@ -206,6 +210,15 @@ public function findJoinCriteriasForTables(LogTable $logTable, $availableLogTabl
+            $otherJoins = $logTable->getWaysToJoinToOtherLogTables();
+            foreach ($otherJoins as $joinTable => $column) {
+                if($availableLogTable->getName() == $joinTable) {
+                    $join = sprintf("%s.%s = %s.%s", $table, $column, $availableLogTable->getName(), $column);
```
There aren't any in the statements created above. Shall we add some everywhere then?

tsteur replied:
Not sure... it may break something (but shouldn't). For the other columns we basically know there won't be any reserved keywords I suppose

sgiehl replied:
I'll add some ticks...

sgiehl committed:
add missing log tables joined using getWaysToJoinToOtherLogTables

tsteur commented on core/DataAccess/LogQueryBuilder/JoinGenerator.php:
```diff
@@ -53,6 +53,10 @@ private function addMissingTablesNeededForJoins()
             if (!$logTable->getColumnToJoinOnIdVisit()) {
                 $tableNameToJoin = $logTable->getLinkTableToBeAbleToJoinOnVisit();
 
+                if (empty($tableNameToJoin) && $logTable->getWaysToJoinToOtherLogTables()) {
```
I'm not 100% sure about this... wouldn't we need to bubble up all ways to join other log tables to see if it can  be joined with the original table eventually? Eg 

`custom_table.userid` => refers to `custom_table2.userid` but this one has no `idvisit` ... then we shouldn't add those tables etc and not assume they can be joined when otherwise only `log_link_visit_action` is present. It may be edge case and should pretty much only happen though when not configuring it properly maybe or when creating log tables that don't make sense in terms of Matomo?

tsteur replied:
or is it not too important because the query will then fail anyway?

sgiehl replied:
I guess the query would fail if something is configured badly. E.g. if a plugin defines dimensions and/or segments for a log table that is not joinable with log_visit or log_action. But I'll have a thought if there's a smart way to check if bubbling up ends in log_visit or log_action

sgiehl replied:
@tsteur I tried starting to implement a way to automatically add all tables up the hierarchy
It gets quite more complicated though. See
https://github.com/matomo-org/matomo/compare/improvejoins...improvejoins2?expand=1

Not sure if it's worth the effort. Otherwise it would only work if a log table defines another way to directly join into `log_visit` or `log_action`

tsteur replied:
OK maybe not too important. I reckon could have been also useful for custom reports to know whether a table can be joined or not in order to eg know which dimensions can be combined etc. I suppose it might not be too important for now.

sgiehl replied:
So how to proceed here now? Adding the table or not?

sgiehl replied:
@tsteur Did some more testing on this. Seems its needed to add required tables up the hierarchy. Otherwise archiving for all reports including conversions fails, as it's not possible to join the custom log table with log_conversion when using a segment for the custom log table. Will merge the other branch here. Would be good if you could do another review then. Will also try to add some more specific tests

tsteur commented on core/DataAccess/LogQueryBuilder/JoinGenerator.php:
```diff
@@ -206,6 +213,15 @@ public function findJoinCriteriasForTables(LogTable $logTable, $availableLogTabl
+            $otherJoins = $logTable->getWaysToJoinToOtherLogTables();
+            foreach ($otherJoins as $joinTable => $column) {
+                if($availableLogTable->getName() == $joinTable) {
+                    $join = sprintf("%s.%s = %s.%s", $table, $column, $availableLogTable->getName(), $column);
```
I'm not 100% sure it looks like this is also used in line 131:

```

                        $table['joinOn'] = $this->findJoinCriteriasForTables($logTable, $availableLogTables);
``` 

when dealing with arrays. but in the above code  in `addMissingTablesNeededForJoins()` we don't add missing tables when it is an array? not sure if you know what I mean, or if it's an issue ... but may need to be looked at and probably needs a test. currently we're adding only needed tables when table is defined as a string `log_xyz` but not through an array `array('table' => 'log_xyz')`.

tsteur commented on core/DataAccess/LogQueryBuilder/JoinGenerator.php:
```diff
@@ -53,6 +53,13 @@ private function addMissingTablesNeededForJoins()
 
+                if (empty($tableNameToJoin) && $logTable->getWaysToJoinToOtherLogTables()) {
+                    foreach ($logTable->getWaysToJoinToOtherLogTables() as $otherLogTable => $column) {
+                        $this->tables->hasJoinedTable($otherLogTable) || $this->tables->addTableToJoin($otherLogTable);
```
can we use a regular if statement for better readability?

sgiehl replied:
sure

sgiehl committed:
automatically add log tables up the hierarchy

sgiehl committed:
code improvements

sgiehl committed:
Adds new ExampleLogTables plugin giving a showcase for custom log tables

sgiehl committed:
specifiy table name in userid archiver to fix query if custom log table joins on user_id column

sgiehl committed:
fix tests

sgiehl committed:
Adds log table that does only indirectly join with log_visit

sgiehl commented:
@tsteur  I've now implemented the automatic join up the hierarchy if it's joinable with `log_visit` or `log_action`. To prove that works as expected I've added a new plugin `ExampleLogTables` which includes two custom log tables. One that joins to `log_visit` using the `user_id` column, and another log table that joins the other custom table using another custom column. Also there is a custom dimension/segment for both log tables.
The plugin includes various system tests to check that all APIs/archiving still work when using a segment of one of the custom log tables.

tsteur commented on core/DataAccess/LogQueryBuilder/JoinGenerator.php:
```diff
@@ -96,6 +109,30 @@ private function addMissingTablesNeededForJoins()
+        $otherTableJoins = $table->getWaysToJoinToOtherLogTables();
+
+        foreach ($otherTableJoins as $logTable => $column) {
+            $this->addMissingTablesForOtherTableJoin($logTable, $tableName);
```
Feel free to ignore but in theory we would maybe only add the tables actually needed. Eg if there's `log_form_foo` and it can be joined with `log_form_bar` and `log_form_baz`, then we would maybe only add `log_form_bar` if this leads us to another log table that we can resolve? It's bit edge case and complicated to solve... the same way as figuring out if a table can be resolved or not... it's a rare feature and so far there's always only one dependency I think so it can be ignored probably until maybe needed or when there's a problem.

sgiehl replied:
So you mean if `getWaysToJoinToOtherLogTables` defines multiple ways to join to different tables?
I'm currently kind of expecting that there is only one possibility defined.

tsteur commented on core/DataAccess/LogQueryBuilder/JoinTables.php:
```diff
@@ -125,6 +159,85 @@ public function sort()
+            return;
+        }
+
+        $this->implicitTableDependencies[$table] = [$dependentTable];
```
Not sure I understand, can't there be multiple dependencies?

sgiehl replied:
In theory there could be multiple, but actually I'm only adding those dependencies, that `$table` is in the list after the static defined ones. Otherwise it can happen that a custom table is sorted before e.g. `log_visit`, which breaks the query. (Removing this would also break the tests I've added)

sgiehl commented:
@tsteur do you have some time for another review? Would be awesome if we could get that merged for 3.9, as I need that for a plugin

tsteur commented on core/DataAccess/LogQueryBuilder/JoinTables.php:
```diff
@@ -125,6 +159,85 @@ public function sort()
+            return true;
+        }
+
+        if ($table->getColumnToJoinOnIdAction()) {
```
is that one actually needed? or is it some edge case?

sgiehl replied:
I'll remove it. Guess I assumed that no one would define multiple methods to allow a join. But actually one could define a way to join on action and on visit. Not sure if that would make sense, but at least it would be possible then.

tsteur commented on core/DataAccess/LogQueryBuilder/JoinTables.php:
```diff
@@ -125,6 +159,85 @@ public function sort()
+            return false;
+        }
+
+        if ($table->getColumnToJoinOnIdVisit()) {
```
same here can a table be joinable on visit and action maybe? not sure I get it

tsteur commented:
Added one more comment otherwise looks fine if all the tests pass.

I reckon there might be edge cases where it doesn't work depending on various combinations of joins for example or when multiple log tables are defined as a way to join etc. But those are not too important right now as we're not using it and can fix the issues as usual once we run into issues.

sgiehl committed:
Allow defining joins on visit and action

sgiehl committed:
update ui files

sgiehl commented:
test failures are unrelated to this PR, so will merge it now.
